design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,432 @@
+# CephX Keyring Rotation
+
+Feature request: https://github.com/rook/rook/issues/15904
+
+## High-level strategy
+
+From an end-user UX view, we can divide keys/keyrings into 2 categories:
+
+### "daemon" keys
+Keys for Ceph daemons that are purely used internally to the Ceph cluster.
+
+Rook can rotate these keys automatically and transparently without risk of disrupting any
+connections to the Ceph cluster beyond Rook's control.
+
+e.g., Admin key(s), mon, mgr, osd, mds, rgw, rbdmirror's internal (non-peer) key
+
+### "Non-daemon" keys
+Any key whose rotation may reasonably require user action beyond Rook API controls.
+
+Because these keys affect non-daemon connections, Rook must make sure the user can initiate rotation
+during a maintenance window. It is imperative that the user can determine when rotation is finished,
+so they know when to update the non-daemon application with the updated key.
+
+e.g., CephClient keys, CephRBDMirror peer keys, CSI keys
+
+Because different types of key rotations may call for different administrator/user workflows, these
+rotations should be able to be initiated independently. For example, a user may have an automated
+workflow to update CephClient A's consumer, but CephClient B's consumer may be a manual process.
+CephClient A and B should allow rotation separately.
+
+CSI keys are a special case that best fits into the non-daemon case. Rotation of CSI keys will affect
+Ceph-CSI's RBD and CephFS volume mounts associated with application pod PVCs. This rotation may
+require administrators to reboot or drain/undrain all nodes with Ceph-CSI PVCs.
+
+## User interfaces
+
+### Key rotation base interface
+
+```yaml
+keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+keyGeneration: <int> # used with KeyGeneration
+```
+
+- `keyRotationPolicy` (string): select a key rotation policy:
+    - `Disabled` (default when unset): don't rotate keys after initial creation
+    - `WithCephVersionUpgrade`: rotate keys when the Ceph version updates
+    - `KeyGeneration`: rotate when the `keyGeneration` input is greater than the current key generation
+
+- `keyGeneration` (int): when `keyRotationPolicy==KeyGeneration`, set this to the desired key
+    generation value. Ignored for other rotation policies. If this is set to greater than the
+    current key generation, relevant keys will be rotated, and the generation value will be updated
+    to this new value (generation values are not necessarily incremental, though that is the
```

Comment:
If we can enforce the incremental generation with the cel rule, no need for this comment.

Reply:
We cannot. I can't find the comment any more, but you identified some corner cases where the CEL rule would prevent users from switching from WithCephVersionUpdate to KeyGeneration options if we wrote the rule.